File: web/src/lib/stages.ts
import type { StageStatus } from "@/lib/types";

export interface StageDef {
  key: string;
  label: string;
  /** gate_key this stage must clear before advancing, or null for no hard gate */
  gateKey: string | null;
}

export const STAGES: StageDef[] = [
  { key: "outcome", label: "Outcome", gateKey: null },
  { key: "intake", label: "Intake", gateKey: null },
  { key: "parse", label: "Parse", gateKey: "parse" },
  { key: "graph", label: "Graph", gateKey: "graph" },
  { key: "explore", label: "Explore", gateKey: null },
  { key: "blueprint", label: "Blueprint", gateKey: "brd_groundedness" },
  { key: "backlog", label: "Backlog", gateKey: "backlog_coverage" },
  { key: "seams", label: "Seams", gateKey: null },
  { key: "plan", label: "Plan", gateKey: "stories_dag" },
  { key: "domain", label: "Domain Design", gateKey: null },
  { key: "design", label: "Design", gateKey: "design_data_ownership" },
  { key: "build", label: "Build", gateKey: "code" },
  { key: "verify", label: "Verify", gateKey: "equivalence" },
];

export interface PhaseDef {
  key: string;
  label: string;
  /** ordered stage keys belonging to this phase */
  stageKeys: string[];
}

/** The three workflow phases, grouping the canonical stages into bands. */
export const PHASES: PhaseDef[] = [
  { key: "understand", label: "Understand", stageKeys: ["outcome", "intake", "parse", "graph", "explore"] },
  { key: "design", label: "Design", stageKeys: ["blueprint", "backlog", "seams", "plan", "domain", "design"] },
  { key: "migrate", label: "Migrate", stageKeys: ["build", "verify"] },
];

/** Global 0-based ordinal of a stage within the full journey. */
export function stageOrdinal(key: string): number {
  return STAGES.findIndex((s) => s.key === key);
}

/** The next stage in the journey after `key`, or null if `key` is last/unknown. */
export function nextStage(key: string): StageDef | null {
  const i = stageOrdinal(key);
  if (i < 0 || i >= STAGES.length - 1) return null;
  return STAGES[i + 1];
}

/**
 * Stages that surface a "next step" advance button in their header. These are
 * the interactive workbench screens where the user drives the flow forward
 * (everything from Graph through Build); the bookend stages are excluded.
 */
export const ADVANCEABLE_STAGES: ReadonlySet<string> = new Set([
  "graph",
  "explore",
  "blueprint",
  "backlog",
  "seams",
  "plan",
  "domain",
  "design",
  "build",
]);

export const STAGE_STATUS_COLOR: Record<StageStatus, string> = {
  pending: "bg-zinc-700 text-zinc-300",
  running: "bg-sky-600 text-white",
  blocked: "bg-amber-600 text-white",
  passed: "bg-emerald-600 text-white",
  failed: "bg-red-600 text-white",
};

export function stageLabel(key: string): string {
  return STAGES.find((s) => s.key === key)?.label ?? key;
}

File: src/cobol_modernizer/controlplane/stages.py
"""The canonical cockpit journey stages (backend source of truth, mirroring
web/src/lib/stages.ts) and their gate-key mapping."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class StageDef:
    key: str
    label: str
    ordinal: int
    gate_key: str | None


JOURNEY_STAGES: list[StageDef] = [
    StageDef("outcome", "Outcome", 0, None),
    StageDef("intake", "Intake", 1, None),
    StageDef("parse", "Parse", 2, "parse"),
    StageDef("graph", "Graph", 3, "graph"),
    StageDef("explore", "Explore", 4, None),
    StageDef("blueprint", "Blueprint", 5, "brd_groundedness"),
    StageDef("seams", "Seams", 6, None),
    StageDef("plan", "Plan", 7, "stories_dag"),
    StageDef("domain", "Domain Design", 8, None),
    StageDef("design", "Design", 9, "design_data_ownership"),
    StageDef("build", "Build", 10, "code"),
    StageDef("verify", "Verify", 11, "equivalence"),
]


def gate_key_for(stage_key: str) -> str | None:
    for s in JOURNEY_STAGES:
        if s.key == stage_key:
            return s.gate_key
    return None
